File: python/rest_producer.py
"""
REST Producer (Python): reads FIXM-inspired JSON, applies S/MIME sign+encrypt,
and POSTs to the consumer endpoint.
Demonstrates SWIM-TIYP-0112 over WS Light with Message Security binding.
"""

import os
import sys
import argparse
import requests
from smime_helper import (
    load_private_key, load_certificate,
    sign, sign_and_encrypt
)


def main():
    parser = argparse.ArgumentParser(description="SWIM S/MIME REST Producer (Python)")
    parser.add_argument("--certs-dir", default=os.path.join(os.path.dirname(__file__), "..", "certs"))
    parser.add_argument("--payload", default=os.path.join(os.path.dirname(__file__), "..", "payload", "sample-flight.json"))
    parser.add_argument("--url", default="http://localhost:8443/receive")
    parser.add_argument("--mode", choices=["sign", "sign-encrypt"], default="sign-encrypt")
    args = parser.parse_args()

    print("=== SWIM S/MIME REST Producer (Python) ===")
    print(f"Mode: {args.mode}")
    print(f"Target: {args.url}")

    with open(args.payload, "rb") as f:
        payload = f.read()
    print(f"Payload loaded: {len(payload)} bytes")

    producer_key = load_private_key(os.path.join(args.certs_dir, "producer.pfx"), "changeit")
    producer_cert = load_certificate(os.path.join(args.certs_dir, "producer.pfx"), "changeit")

    if args.mode == "sign-encrypt":
        consumer_cert = load_certificate(os.path.join(args.certs_dir, "consumer.crt"))
        smime_data, content_type = sign_and_encrypt(
            payload, "application/json",
            producer_key, producer_cert, consumer_cert)
        print("Message signed and encrypted")
    else:
        smime_data, content_type = sign(
            payload, "application/json",
            producer_key, producer_cert)
        print("Message signed")

    print(f"S/MIME Content-Type: {content_type}")
    print(f"S/MIME message size: {len(smime_data)} bytes")

    # Send via HTTP POST per SWIM-TIYP-0112 constraints
    headers = {
        "Content-Type": content_type,
        "MIME-Version": "1.0",
    }

    response = requests.post(args.url, data=smime_data, headers=headers)
    print(f"Response status: {response.status_code}")
    print(f"Response body: {response.text}")


if __name__ == "__main__":
    main()

File: java/src/main/java/com/swim/smime/RestProducer.java
package com.swim.smime;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;

/**
 * REST Producer: reads FIXM-inspired JSON, applies S/MIME signing (or sign+encrypt),
 * and POSTs to the consumer endpoint.
 * Demonstrates SWIM-TIYP-0112 over WS Light with Message Security binding.
 */
public class RestProducer {

    public static void main(String[] args) throws Exception {
        String certsDir = args.length > 0 ? args[0] : "../certs";
        String payloadFile = args.length > 1 ? args[1] : "../payload/sample-flight.json";
        String consumerUrl = args.length > 2 ? args[2] : "http://localhost:5000/receive";
        String mode = args.length > 3 ? args[3] : "sign"; // "sign" or "sign-encrypt"

        System.out.println("=== SWIM S/MIME REST Producer (Java) ===");
        System.out.println("Mode: " + mode);
        System.out.println("Target: " + consumerUrl);

        // Load payload
        byte[] payload = Files.readAllBytes(Path.of(payloadFile));
        System.out.println("Payload loaded: " + payload.length + " bytes");

        // Load producer credentials (for signing)
        PrivateKey producerKey = SmimeHelper.loadPrivateKey(
                Path.of(certsDir, "producer.pfx").toString(), "changeit");
        X509Certificate producerCert = SmimeHelper.loadCertificateFromPfx(
                Path.of(certsDir, "producer.pfx").toString(), "changeit");

        SmimeHelper.SmimeResult result;

        if ("sign-encrypt".equals(mode)) {
            // Load consumer certificate (for encryption)
            X509Certificate consumerCert = SmimeHelper.loadCertificate(
                    Path.of(certsDir, "consumer.crt").toString());
            result = SmimeHelper.signAndEncrypt(payload, "application/json",
                    producerKey, producerCert, consumerCert);
            System.out.println("Message signed and encrypted");
        } else {
            result = SmimeHelper.sign(payload, "application/json",
                    producerKey, producerCert);
            System.out.println("Message signed");
        }

        // Unfold MIME header line continuations (tabs/spaces after line breaks)
        String contentType = result.contentType().replaceAll("\\r?\\n[ \\t]+", " ");
        System.out.println("S/MIME Content-Type: " + contentType);
        System.out.println("S/MIME message size: " + result.data().length + " bytes");

        // Send via HTTP POST per SWIM-TIYP-0112 constraints
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(consumerUrl))
                .header("Content-Type", contentType)
                .header("MIME-Version", "1.0")
                .POST(HttpRequest.BodyPublishers.ofByteArray(result.data()))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

        System.out.println("Response status: " + response.statusCode());
        System.out.println("Response body: " + response.body());
    }
}
